DEVX-129 add doc for cart discount sync (#1030)
commercetools/commercetools-sync-java: docs/sdk2/usage/CART_DISCOUNT_SYNC.md

When the reference key is provided with a `ResourceIdentifier`, the sync will resolve the resource with the given key and use the ID of the found resource to create or update a reference.
Therefore, in order to resolve the actual ids of those references in the sync process, `ResourceIdentifier`s with their `key`s have to be supplied. 

|Reference Field|Type|
|:---|:---|
| `custom.type` | ResourceIdentifier to a Type |  

> Note that a reference without the key field will be considered as an existing resource on the target commercetools project and the library will issue an update/create an API request without reference resolution.

##### Syncing from a commercetools project

When syncing from a source commercetools project, you can use [`toCartDiscountDrafts`](#todo)
method that transforms(resolves by querying and caching key-id pairs) and maps from a `CartDiscount` to `CartDiscountDraft` using cache in order to make them ready for reference resolution by the sync, for example: 

````java
// Build a ByProjectKeyCartDiscountsGet for fetching cart discounts from a source CTP project without any references expanded for the sync:

> final CartDiscountQuery cartDiscountQuery = client.categories().get();

Reviewer: ```suggestion
final ByProjectKeyCartDiscountsGet byProjectKeyCartDiscountsGet = client.categories().get();
```

Author: https://github.com/commercetools/commercetools-sync-java/pull/1030/commits/d458876d5189a38b3c0c0ad325b2de1292712ad4

Author: https://github.com/commercetools/commercetools-sync-java/pull/1030/commits/dd4d6bc9f113b0d4e2335d1e9790827537133834

CompletableFuture<List<CartDiscountDraft>> cartDiscountDrafts = CartDiscountTransformUtils.toCartDiscountDrafts(client, referenceIdToKeyCache, cartDiscounts);
````

##### Syncing from an external resource

- When syncing from an external resource, `ResourceIdentifier`s with their `key`s have to be supplied as following example:

```` java
final CartDiscountDraft cartDiscountDraft = CartDiscountDraftBuilder.of()
            .cartPredicate("cartPredicate")
            .name(LocalizedString.ofEnglish("foo"))
            .requiresDiscountCode(true)
            .sortOrder("0.1")
            .target(CartDiscountTargetBuilder::shippingBuilder)
            .value(CartDiscountValueAbsoluteDraftBuilder.of()
                    .money(
                            MoneyBuilder.of()
                                    .centAmount(10L)
                                    .currencyCode(DefaultCurrencyUnits.EUR.getCurrencyCode())
                                    .build())
                    .build())
            .key("cart-discount-key")

>             .custom(CustomFieldsDraft.ofTypeKeyAndJson("type-key", emptyMap())) // note that custom type provided with key

Reviewer: The method `CustomFieldsDraft.ofTypeKeyAndJson` isn't valid vor SDK v2. Its CustomFieldsDraftBuilder.of()

Author: https://github.com/commercetools/commercetools-sync-java/pull/1030/commits/d458876d5189a38b3c0c0ad325b2de1292712ad4

            .build();
````

#### SyncOptions

After the `ProjectApiRoot` is set up, a `CartDiscountSyncOptions` should be built as follows:

````java
// instantiating a CartDiscountSyncOptions
final CartDiscountSyncOptions cartDiscountSyncOptions = CartDiscountSyncOptionsBuilder.of(projectApiRoot).build();
````

`SyncOptions` is an object which provides a place for users to add certain configurations to customize the sync process.
Available configurations:

##### errorCallback
A callback that is called whenever an error event occurs during the sync process. Each resource executes its own 
error-callback. When the sync process of a particular resource runs successfully, it is not triggered. It contains the 
following context about the error-event:

* sync exception
* cart discount draft from the source
* cart discount of the target project (only provided if an existing cart discount could be found)
* the update-actions, which failed (only provided if an existing cart discount could be found)

````java
 final Logger logger = LoggerFactory.getLogger(CartDiscountSync.class);
 final CartDiscountSyncOptions cartDiscountSyncOptions = CartDiscountSyncOptionsBuilder

>          .of(sphereClient)

Reviewer: ```suggestion
         .of(projectApiRoot)
```

Author: https://github.com/commercetools/commercetools-sync-java/pull/1030/commits/d458876d5189a38b3c0c0ad325b2de1292712ad4

         .errorCallback((syncException, draft, cartDiscount, updateActions) -> 
            logger.error(new SyncException("My customized message"), syncException)).build();
````
    
##### warningCallback
A callback is called whenever a warning event occurs during the sync process. Each resource executes its own 
warning-callback. When the sync process of a particular resource runs successfully, it is not triggered. It contains the 
following context about the warning message:

* sync exception
* cart discount draft from the source 
* cart discount of the target project (only provided if an existing cart discount could be found)

````java
 final Logger logger = LoggerFactory.getLogger(CartDiscountSync.class);
 final CartDiscountSyncOptions cartDiscountSyncOptions = CartDiscountSyncOptionsBuilder

>          .of(sphereClient)

Reviewer: ```suggestion
         .of(projectApiRoot)
```

Author: https://github.com/commercetools/commercetools-sync-java/pull/1030/commits/d458876d5189a38b3c0c0ad325b2de1292712ad4

            logger.warn(new SyncException("My customized message"), syncException)).build();
````

##### beforeUpdateCallback
During the sync process, if a target cart discount and a cart discount draft are matched, this callback can be used to 
intercept the **_update_** request just before it is sent to the commercetools platform. This allows the user to modify 
update actions array with custom actions or discard unwanted actions. The callback provides the following information :
 
 * cart discount draft from the source
 * cart discount from the target project
 * update actions that were calculated after comparing both

````java
final TriFunction<
        List<CartDiscountUpdateAction>, CartDiscountDraft, CartDiscount, List<CartDiscountUpdateAction>> 
            beforeUpdateCartDiscountCallback =
            (updateActions, newCartDiscountDraft, oldCartDiscount) ->  updateActions.stream()
                    .filter(updateAction -> !(updateAction instanceof CartDiscountChangeCartPredicateAction))
                    .collect(Collectors.toList());
                        
final CartDiscountSyncOptions cartDiscountSyncOptions = 

>         CartDiscountSyncOptionsBuilder.of(sphereClient).beforeUpdateCallback(beforeUpdateCartDiscountCallback).build();

Reviewer: ```suggestion
        CartDiscountSyncOptionsBuilder.of(projectApiRoot).beforeUpdateCallback(beforeUpdateCartDiscountCallback).build();
```

Author: https://github.com/commercetools/commercetools-sync-java/pull/1030/commits/d458876d5189a38b3c0c0ad325b2de1292712ad4

````java                         
final CartDiscountSyncOptions cartDiscountSyncOptions = 
         CartDiscountSyncOptionsBuilder.of(projectApiRoot).batchSize(30).build();
````

##### cacheSize
In the service classes of the commercetools-sync-java library, we have implemented an in-memory [LRU cache](https://en.wikipedia.org/wiki/Cache_replacement_policies#Least_recently_used_(LRU)) to store a map used for the reference resolution of the library.
The cache reduces the reference resolution based calls to the commercetools API as the required fields of a resource will be fetched only one time. These cached fields then might be used by another resource referencing the already resolved resource instead of fetching from commercetools API. It turns out, having the in-memory LRU cache will improve the overall performance of the sync library and commercetools API.
which will improve the overall performance of the sync and commercetools API.

Playing with this option can change the memory usage of the library. If it is not set, the default cache size is `10.000` for cart discount sync.
````java
final CartDiscountSyncOptions cartDiscountSyncOptions = 

>          CartDiscountSyncOptionsBuilder.of(sphereClient).cacheSize(5000).build(); 

Reviewer: ```suggestion
         CartDiscountSyncOptionsBuilder.of(projectApiRoot).cacheSize(5000).build(); 
```

Author: https://github.com/commercetools/commercetools-sync-java/pull/1030/commits/d458876d5189a38b3c0c0ad325b2de1292712ad4

any HTTP client module. The default one is `commercetools-http-client`.

```xml
 <!-- Sample maven pom.xml -->
 <properties>
     <commercetools.version>LATEST</commercetools.version>
 </properties>

 <dependencies>
     <dependency>
       <groupId>com.commercetools.sdk</groupId>
       <artifactId>commercetools-http-client</artifactId>
       <version>${commercetools.version}</version>
     </dependency>
     <dependency>
       <groupId>com.commercetools.sdk</groupId>
       <artifactId>commercetools-sdk-java-api</artifactId>
       <version>${commercetools.version}</version>
     </dependency>
 </dependencies>

```

### Client configuration and creation

For client creation use [ClientConfigurationUtils](#todo) which apply the best practices for `ProjectApiRoot` creation.
If you have custom requirements for the client creation make sure to replace `SphereClientFactory` with `ApiRootBuilder` as described in this [Migration Document](https://docs.commercetools.com/sdk/java-sdk-migrate#client-configuration-and-creation).

### Signature of CartDiscountSyncOptions

> As models and update actions have changed in the JVM-SDK-V2 the signature of SyncOptions is different. It's constructor now takes a `ProjectApiRoot` as first argument. The callback functions are signed with `CartDiscount`, `CartDiscountDraft` and `CartDiscountUpdateAction` from `package com.commercetools.api.models.category.*`

Reviewer: ```suggestion
As models and update actions have changed in the JVM-SDK-V2 the signature of SyncOptions is different. It's constructor now takes a `ProjectApiRoot` as first argument. The callback functions are signed with `CartDiscount`, `CartDiscountDraft` and `CartDiscountUpdateAction` from `package com.commercetools.api.models.cartDiscount.*`
```

Author: https://github.com/commercetools/commercetools-sync-java/pull/1030/commits/d458876d5189a38b3c0c0ad325b2de1292712ad4

> Note: Type `UpdateAction<CartDiscount>` has changed to `CartDiscountUpdateAction`. Make sure you create and supply a specific CartDiscountUpdateAction in `beforeUpdateCallback`. For that you can use the [library-utilities](#todo) or use a JVM-SDK builder ([see also](https://docs.commercetools.com/sdk/java-sdk-migrate#update-resources)):

```java

> // Example: Create a category update action to change name taking the 'newName' of the categoryDraft

Reviewer: ```suggestion
// Example: Create a cartDiscount update action to change name taking the 'newName' of the cartDiscountDraft
```

Author: https://github.com/commercetools/commercetools-sync-java/pull/1030/commits/d458876d5189a38b3c0c0ad325b2de1292712ad4

The V2 SDK do not have inheritance for `DraftBuilder` classes but the differences are minor and you can replace it easily. Here's an example:

```java
// CartDiscountDraftBuilder in v1 takes parameters 'name', 'cartPredicate', 'value', 'target', 'sortOrder', 'requiresDiscountCode' 
final CartDiscountDraft cartDiscountDraft =
        CartDiscountDraftBuilder.of(
              ofEnglish("name"),
              "1 = 1",
              CartDiscountValue.ofAbsolute(MoneyImpl.of(20, EUR)),
              "sku = \"0123456789\" or sku = \"0246891213\"",
              "0.2",
              false)
          .key("key")
          .active(false)
          .build()

// CartDiscountDraftBuilder in v2
final CartDiscountDraft cartDiscountDraft =
        CartDiscountDraftBuilder.of()
            .name(ofEnglish("name"))
            .cartPredicate("1 = 1")
            .value(
              CartDiscountValueAbsoluteDraftBuilder.of()
                .money(
                    CentPrecisionMoneyBuilder.of()
                      .centAmount(20L)
                      .fractionDigits(0)

>                       .currencyCode(EUR.getCurrencyCode())

Reviewer: ```suggestion
                      .currencyCode(DefaultCurrencyUnits.EUR.getCurrencyCode())
```

Author: https://github.com/commercetools/commercetools-sync-java/pull/1030/commits/d458876d5189a38b3c0c0ad325b2de1292712ad4
